### DISCUSSION TITLE
Refactor MDX to output JS directly

### DISCUSSION THREAD

**Discussion by bluwy** [withastro/roadmap#586] [Proposal]
### Body

# Summary

Refactor `@astrojs/mdx` to output JS instead of JSX. The JS code uses Astro’s rendering API similar to the output code of an Astro file.

# Background & Motivation

1. MDX JSX AST is huge and slows down Rollup.
2. MDX JSX goes through a lot of preprocessing to be useable by Astro.
3. MDX JSX rendering isn’t optimal and re-implements a lot of Astro’s rendering API.
4. Astro needs to support JSX that’s only used by `@astrojs/mdx`.

By compiling directly to JS, we can solve the issues above. I’ve made an [experimental branch](https://github.com/withastro/astro/tree/wip-mdx-to-astro-js) that implements it, I think it can be polished up and replace the current flow.

I've also sent https://github.com/withastro/astro/pull/7151 that solves no1 and is probably good enough in many case. But metaphorically, it only patches a problem we had, "compiling to JS" would be treating the symptom directly.

# Notes

1. The current MDX flow goes: `.mdx -> mdast -> hast -> esast -> .jsx -> .js`
2. The proposed flow goes: `.mdx -> mdast -> hast -> .js`
3. Astro-flavoured markdown goes: `.md -> .astro -> .ts -> .js`
4. In Astro 3.0, we can remove Astro’s JSX handling and break off the old code.
5. I think this decision also depends on how Markdoc progresses (I personally prefer the MDX syntax 😬)

**Comment by matthewp** [comment]
I like the idea generally speaking; more performant would be better. My main concern would be if this *increases* maintenance burden by doing things in a different way that in a typical MDX (React) app? Or are we already doing things differently enough because of our custom JSX. 

Should this change be noticeable by the user (aside from perf)?

**Reply by bluwy** [reply]
The maintenance would move from our JSX processing and runtime, to a rehype plugin, so I think it'll be even less maintenance. It would take more effort at the start to ensure previous compatibility though, but I think it'll pay off in the long run.

I don't think we're doing it differently than a MDX + React setup currently, as we're playing by it's rules (emit JSX + render with a JSX runtime). My proposal would deviate from that rule, but it'll only be an implementation detail. The end user shouldn't notice anything different. (The idea is similar to how Solid compiles JSX differently)

**Comment by bluwy** [comment]
Digging more into AFMD today, I’m pretty convinced now I’m creating AFMD 2.0 here 😅 I could’ve also written a rehype plugin that converts to raw Astro code instead and it’ll be it (because then we’ll be able to get all the Astro features in MDX).

**Comment by ematipico** [comment]
That's essentially the same idea I had in mind when I saw the AST emitted by MDX. There's a TON of optimisation that we can do before passing the final code to the runtime.

I like the proposal!
